### PR TITLE
Kotlin: extract function references using compiler-generated adapters

### PR DESCRIPTION
**Pull request by smowton**

### CODE REVIEW

File: java/kotlin-extractor/src/main/kotlin/KotlinFileExtractor.kt [hunk 1]
```diff
@@ -4487,7 +4493,10 @@ open class KotlinFileExtractor(
                 val baseClass = pluginContext.referenceClass(FqName("kotlin.jvm.internal.FunctionReference"))?.owner?.typeWith()
                     ?: pluginContext.irBuiltIns.anyType
 
-                val classId = extractGeneratedClass(ids, listOf(baseClass, fnInterfaceType), locId, functionReferenceExpr, declarationParent)
+                val classId = extractGeneratedClass(ids, listOf(baseClass, fnInterfaceType), locId, functionReferenceExpr, declarationParent, { it.valueParameters.size == 1 }) {
+                    // The argument to FunctionReference's constructor is the function arity.
```
Comment (tamasvajk) [inline java/kotlin-extractor/src/main/kotlin/KotlinFileExtractor.kt:4497]: Is this also the case for big arity functions?

**Review by tamasvajk — DISMISSED** [review]
Looks reasonable to me. We could change to use the `reference.symbol` if all our tests are passing with that. (Slightly odd that `reference.reflectionTarget` is nullable, while `symbol` is not.)

File: java/kotlin-extractor/src/main/kotlin/KotlinFileExtractor.kt (hunk 1 above)
Comment (smowton) [reply java/kotlin-extractor/src/main/kotlin/KotlinFileExtractor.kt:4497]: The base class is `kotlin.jvm.internal.FunctionReference` regardless of the arity

**Review by tamasvajk — APPROVED** [review]

**Comment by smowton** [comment]
@tamasvajk rebased, please re-approve?

**Review by tamasvajk — APPROVED** [review]

**Pull request merged**